### PR TITLE
CI: Fix Docker Hub image build & push

### PR DESCRIPTION
**Pull request by snpranav**
Due to issue #848, I don't think we can rely on the Docker Hub auto-builds to have the latest image. Hence, I wrote a workflow to build and push the docker image to the Github Container Registry ([ghcr.io](https://ghcr.io/)).

For it to work, a personal access token needs to be added to the secrets of the repository to be able to authenticate with the docker registry. The secret must be stored with the name `CR_PAT`.

Once this is set up, the community can maintain the workflow to build and push the latest docker images.

### CODE REVIEW

**Comment by ronjouch** [comment]
Hi @snpranav, thanks for the initiative! Jia (original author of Nativefier, and managing our publishing) is going to look at it.

**Comment by jiahaog** [comment]
Thanks for the contribution! It's unfortunate that the automated builds on DockerHub are just not working for some unknown reason :(

Just some preliminary feedback here – do you think it would be better to add this as another step to the [CI workflow](https://github.com/jiahaog/nativefier/blob/master/.github/workflows/ci.yml)? I'm not really about the conventions of GitHub actions, but it does seem like it would be useful to do this publishing step only after the tests run and have passed.

Also, can we push the images to DockerHub instead since users are already using that repository? Happy to provide the credentials if so 🙂

**Comment by snpranav** [comment]
Hi @jiahaog,
Thanks for reviewing the pull request.

> Just some preliminary feedback here – do you think it would be better to add this as another step to the [CI workflow](https://github.com/jiahaog/nativefier/blob/master/.github/workflows/ci.yml)? I'm not really about the conventions of GitHub actions, but it does seem like it would be useful to do this publishing step only after the tests run and have passed.

Haha. I'm used to GitLab CI, so GitHub actions are quite new to me also :smile:. Nevertheless, I have updated the docker build and the push process to happen in the [CI file](https://github.com/jiahaog/nativefier/blob/master/.github/workflows/ci.yml) after the test job passes.

The reason the docker build test in the pull request is failing is that two secrets in the Github repository are missing.
You need to add the following secrets to the GitHub repository.
1. `DOCKERHUB_USERNAME` - Set it to your Docker Hub username (in this case I'm guessing it would just be `jiahaog`).
2. `DOCKERHUB_TOKEN` - Set it to the personal access token from Docker Hub.

Once you add these secrets the CI pipeline will run the tests, build the docker image and push it to Docker Hub with the latest tag and the short commit SHA tag.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -31,3 +31,34 @@ jobs:
...
+    env:
+      IMAGE_ID: ${{ secrets.DOCKERHUB_USERNAME }}/nativefier
+    
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Build image
+        run: docker build . --file Dockerfile --tag $IMAGE_ID:latest
```
Comment (jiahaog) [inline .github/workflows/ci.yml:48]:
```suggestion
        run: docker build . --file Dockerfile --tag "$IMAGE_ID:latest"
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -31,3 +31,34 @@ jobs:
...
+      - name: Login to Docker Hub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+          
+      - name: Tag and Push
+        run: |
+          # Tag docker image the commit SHA
+          docker tag $IMAGE_ID:latest $IMAGE_ID:${GITHUB_SHA::8}
```
Comment (jiahaog) [inline .github/workflows/ci.yml:59]:
```suggestion
          docker tag "$IMAGE_ID:latest" "$IMAGE_ID:${GITHUB_SHA::8}"
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -31,3 +31,34 @@ jobs:
...
+          
+          # Push the images to the registry
+          docker push $IMAGE_ID:latest
+          docker push $IMAGE_ID:${GITHUB_SHA::8}
```
Comment (jiahaog) [inline .github/workflows/ci.yml:63]:
```suggestion
          docker push "$IMAGE_ID:latest"
          docker push "$IMAGE_ID:${GITHUB_SHA::8}"
```

**Review by jiahaog — APPROVED** [review]
Sorry for the delay, I've been a bit busy. But thanks a lot for the contribution! Just a few nits about quoting the shell variables, and we can merge this :)

**Comment by snpranav** [comment]
Just incorporated the suggestions into this PR. Thanks for approving the PR and creating this repository @jiahaog.
Thanks @ronjouch for your amazing contribution and maintaining this repository. :smile: :partying_face: 

@jiahaog, I don't think you have put in your credentials for DockerHub in the secrets as specified in my previous commend. The workflow is failing in the PR due to this.

**Comment by jiahaog** [comment]
Oh, I think you won't be able to access the secrets anyway in a pull request - see https://docs.github.com/en/actions/reference/encrypted-secrets#using-encrypted-secrets-in-a-workflow. I was intending to add the secrets then merge this to test if the workflow works, I'm pretty confident you did a good job :)

However... as I was looking for the secrets just now, it seems like Docker Hub has magically started working again... 
Screen Shot 2021-01-27 at 22 17 15

I think they might have fixed this on their end. I'd rather let Docker Hub take care of this instead of having to maintain our workflows, would you be alright if we keep this around for a couple of weeks and see how it goes? If Docker Hub stops working again we should definitely merge this pull request. So sorry for the extra work you had to do.

**Comment by ronjouch** [comment]
> I think they might have fixed this on their end. I'd rather let Docker Hub take care of this instead of having to maintain our workflows, would you be alright if we keep this around for a couple of weeks and see how it goes? If Docker Hub stops working again we should definitely merge this pull request. So sorry for the extra work you had to do.

I confirm that Docker seems to work again now. I just made a new release, and I see Docker Hub picked up on it.

With that, like Jia said, @snpranav thanks for all your work 🙂.

Feel free to chime in here or at https://github.com/nativefier/nativefier/issues/848 if Docker Hub starts lagging again.

**Comment by snpranav** [comment]
@jiahaog @ronjouch As mentioned in #848, the docker auto builds have stopped again. I would advocate for reconsidering accepting this pull request as fixing the docker hub auto builds always takes a really long time (this time it has been 5 months and still not fixed).

**Review by ronjouch — CHANGES_REQUESTED** [review]
> As mentioned in #848, the docker auto builds have stopped again. I would advocate for reconsidering accepting this pull request as fixing the docker hub auto builds always takes a really long time (this time it has been 5 months and still not fixed).

@snpranav yup, agreed. We gave Docker Hub's auto-build a try and it would have been the ideal solution, but it's still too broken. Oh well.

This PR needs a rework, though. It's set to happen in CI, for *all builds* including PR builds! This means all WIP PRs running CI will push to Docker, and we certainly do not want that :scream: . Please move your new job to `publish.yml`, not `ci.yml`

**Comment by snpranav** [comment]
@ronjouch Good catch :sweat_smile: , I just moved the docker build actions from `ci.yml` file to the `publish.yml`. Let me know if you see any other issues with the docker build action :)

**Comment by snpranav** [comment]
Hi @ronjouch , I implemented the changes you had requested in this PR a few weeks ago. Could you please tell me if there's any status update on this PR?

Reference - #848

**Review by ronjouch — APPROVED** [review]

**Comment by ronjouch** [comment]
@snpranav confirmed working, thanks! https://github.com/nativefier/nativefier/issues/848#issuecomment-1006706120

**Pull request merged**